### PR TITLE
chore(gatsby): convert redux/reducers/pages to typescript

### PR DESCRIPTION
**Pull request by kpfromer**
## Description

This will convert the `redux/reducers/pages` to TypeScript. I also changed the test file to TypeScript.

## Related Issues

Related to #21995

### CODE REVIEW

File: packages/gatsby/src/redux/reducers/pages.ts [hunk 1]
```diff
@@ -1,12 +1,21 @@
...
+  state: IGatsbyState["pages"] = new Map(),
+  action: IDeleteCacheAction | ICreatePageAction | IDeletePageAction
+): IGatsbyState["pages"] => {
   switch (action.type) {
     case `DELETE_CACHE`:
       return new Map()
     case `CREATE_PAGE`: {
       action.payload.component = normalize(action.payload.component)
-      if (!action.plugin && !action.plugin.name) {
+      if (!action.plugin || !action.plugin.name) {
```
Comment (LekoArts) [inline packages/gatsby/src/redux/reducers/pages.ts:18]: This is changing behavior, why was that changed?

Comment (kpfromer) [reply packages/gatsby/src/redux/reducers/pages.ts:18]: With the original code, the `!action.plugin.name` was short-circuited and TypeScript was saying that `action.plugin.name not assignable to parameter of type never`.

Comment (LekoArts) [reply packages/gatsby/src/redux/reducers/pages.ts:18]: Ah, yeah, sure! We can use optional chaining here, so `!action.plugin?.name`

Comment (kpfromer) [reply packages/gatsby/src/redux/reducers/pages.ts:18]: So it would look something like `if (!action.plugin && !action.plugin?.name)` ?

Comment (LekoArts) [reply packages/gatsby/src/redux/reducers/pages.ts:18]: It would be `if(!action.plugin?.name)` - you can read more here: https://github.com/tc39/proposal-optional-chaining#overview-and-motivation

Comment (kpfromer) [reply packages/gatsby/src/redux/reducers/pages.ts:18]: Whoa! This is really cool. Thanks for the information, the commit should be up soon.

**Comment by kpfromer** [comment]
I messed up my fork repo and will create a new pull request with the merge conflicts fixed.

**Comment by pvdz** [comment]
@kpfromer can you please let us know if you still intend to do this? In that case can you link to this PR and then close this PR in favor of the other one? Thanks! :purple_heart:

**Comment by kpfromer** [comment]
@pvdz Sorry for the late response, I have been quite busy. I will try to get this done by the end of the weekend (July 12).
